Add Shardy support for Torch-XLA

hshahTT opened the pull request:
This PR implements the necessary changes to support the Shardy dialect within Torch-XLA (relevant issue: https://github.com/pytorch/xla/issues/9348):

1. Adding support for V2 HLO sharding within the `OpSharding` and `XlaShardingSpec` classes  (since Shardy doesn't support the V1 shardings that are currently implemented).
2. Add the OpenXLA [`addStablehloImportPipeline()`](https://github.com/openxla/xla/blob/0cead9fb0f6b5a3effbfc90858640f2234e1c76c/xla/service/spmd/shardy/stablehlo_round_trip/stablehlo_import.cc#L762) pass that performs the SHLO to Shardy conversion.
3. This is protected by the `"CONVERT_SHLO_TO_SHARDY"` environment variable.

hshahTT commented:
Hi, could I get some eyes on this please?

ZixuanJiang commented on torch_xla/csrc/xla_sharding_util.cpp:
```diff
@@ -218,6 +218,23 @@ bool ShardingUtil::EqualOpShardings(const xla::OpSharding& a,
+  auto transpose_perm_vec = transpose_perm.cast<std::vector<int>>();
+  std::vector<xla::OpSharding::Type> subgroup_types;
+  if (dims_vec.size() > transpose_perm.size()) {
+    subgroup_types.push_back(xla::OpSharding::REPLICATED);
```
`reshape_dims` and `transpose_perm` only decides the device list, which is unrelated to subgroup_types.

Subgroup types should be another input argument of this function.

hshahTT replied:
Moved the subgroup_types logic into `xla_sharding.py` and added that as an input argument instead.

ZixuanJiang commented on torch_xla/csrc/xla_sharding_util.cpp:
```diff
@@ -218,6 +218,23 @@ bool ShardingUtil::EqualOpShardings(const xla::OpSharding& a,
 
+xla::OpSharding ShardingUtil::CreateIotaOpSharding(
+    const py::list& dims, const py::list& reshape_dims,
+    const py::list& transpose_perm) {
```
Add CHECK_EQ(reshape_dims.size(), transpose_perm.size())

hshahTT replied:
Done

ZixuanJiang commented on torch_xla/distributed/spmd/xla_sharding.py:
```diff
@@ -142,6 +152,63 @@ def _get_op_sharding_args(self, partition_spec: PartitionSpec):
     return tile_assignment, group_assignment, replication_groups, sharding_type
 
+  @functools.lru_cache(maxsize=None)
+  def _get_op_sharding_args_v2(self, partition_spec: PartitionSpec):
```
In [this function](https://github.com/openxla/xla/blob/main/xla/service/spmd/shardy/stablehlo_round_trip/export_shardings.h#L34), we convert Shardy sharding (which is based on mesh and axes) into HloShardingV2. Could you please refer to it?

hshahTT commented:
Thanks @ZixuanJiang for your review! The link you gave me to the OpenXLA implementation and your comment [here](https://github.com/pytorch/xla/issues/9348#issuecomment-3192952206) was very useful.

Based on your comments I made the following changes:
- **xla_sharding.py:**
    - Modified the logic in `_get_op_sharding_args_v2()` to match the [convertToHloSharding()](https://github.com/openxla/xla/blob/2e237f17fbafd2571b0e655c85b9137185e8c99a/xla/service/spmd/shardy/stablehlo_round_trip/export_shardings.cc#L288) function within OpenXLA

- **test_xla_sharding.py:**
    - Modified the existing test cases to test the V2 sharding logic when the `CONVERT_SHLO_TO_SHARDY` environment variable is set.
    - The way it is currently setup requires you to set the environment variable before running the test (i.e., by running `CONVERT_SHLO_TO_SHARDY=1 python test_xla_sharding.py` in whatever Bash script actually calls it in CI). 
    - Another way is to parameterize the testing class itself based on whether the env var should be set or not (meaning we run all the tests once with the env variable unset and then again with it set). I can do this but may need to add the [parameterized](https://pypi.org/project/parameterized/) pip module as a test dependency to do it cleanly.

- **debugging.py:**
    - Added support for V2 shardings within the `visualize_tensor_sharding()` debugging function by converting them into V1 shardings first (which are already supported) via the `construct_v1_sharding_str()` function.
    - I added a `_get_xla_op_sharding_v2_params` Pybind function inside `init_python_bindings.cpp` that takes a tensor and returns all the V2 sharding parameters needed to represent it: `tile_assignment_dims, reshape_dims, transpose_perm, is_last_tile_dim_replicate`.
    - Another way I can get those params is by just parsing the V2 sharding string directly, since it already has the format:
      ```
      {devices=[tile_assignment_dims]<=[reshape_dims]T(transpose_perm) last_tile_dim_replicate}
      ```
      but I thought that the Pybind function would be more readable.

- **test_spmd_debugging.py:** Added the `ConvertV2ShardingToV1Test` testing class to test the `construct_v1_sharding_str()` function.

I tested everything on a Cloud v4-8 TPU courtesy of Google's TPU Research Cloud program. Please let me know if anything else needs to be done!

hshahTT commented:
Saw the failing build on CI and added a fix for it: https://github.com/pytorch/xla/pull/9541/commits/d1e0b1707de161818fa9d699b9c356fc5ab189f5. Could someone please re-run it?

hshahTT commented:
Thanks for re-running the CI! I fixed the errors in `ConvertV2ShardingToV1Test` in [this](https://github.com/pytorch/xla/pull/9541/commits/ffb228de268b452c36d2f1f16055462c605b1b99) commit. However, I wasn't able to reproduce the failure in [test/tpu/run_training_tests.sh](https://github.com/pytorch/xla/actions/runs/18228844186/job/52068529567?pr=9541) on a v4-8 TPU. It also seems unrelated to my changes in this PR since the `CONVERT_SHLO_TO_SHARDY` flag wasn't enabled for that test. Does anyone what might be happening there?
